### ISSUE TITLE
Training VQGAN on TPU

### ISSUE THREAD

**Issue by tmabraham** [tgisaturday/dalle-lightning#1]
I am trying to train a VQGAN on an 8-core TPU. Have you successfully trained a VQGAN on a single TPU using this codebase?

**Comment by tgisaturday** [comment]
Yes. I've tested VQGAN with V3-8, V3-32, and V3-64. Training fails on larger Pods.

**Comment by tgisaturday** [comment]
@tmabraham One concern is that I've recently modified VQGANs with newly updated version. I'm testing them right now.

**Comment by tmabraham** [comment]
@tgisaturday After modifying the taming-transformers repo and also trying your repository, I have noticed that training on V3-8 is too slow and profiling indicates this is because of an unsupported operation (sgn_out). Have you solved this issue?

**Comment by tgisaturday** [comment]
@tmabraham Yes it is slow. takes about 7s per iter. Did you find out where sgn_out is located? I guess we can ask for operation lowering.

**Comment by tgisaturday** [comment]
@tmabraham And also .detach() itself is huge overload for TPU because it requires recompiling everytime. Changing the loss with masking is known to be helpful.

**Comment by tgisaturday** [comment]
@tmabraham I've just refactored vqgan code and it is now fast.
Use `--model` flag to train different models with `train_vae.py`.

For v3-8 single node tpu-vm:
vqvae: 1.02s
vqvae2: 1.37s
vqgan: 3.06s
gvqvae: 1.42s (Gumbel-softmax VQ)
gvqgan: 3.57s
evqvae: 1.22s (EMA VQ)
evqgan: 3.25s

per iteration.
